### PR TITLE
Support IAccessible2 labelled-by relation

### LINKED ISSUE #17436: IAccessible2 "labelled-by" property not supported
IAccessible2 has an IA2_RELATION_LABELLED_BY relation type, see https://accessibility.linuxfoundation.org/a11yspecs/ia2/docs/html/group__grp_relations.html#ga7bbace7ffb57476b75d621af2e27d1ff

However, that one is not taken into account by NVDA's "labeledBy" property for objects, as can be seen for example with LibreOffice that implements handling of the IAccessible2 property.

### Steps to reproduce:

1. start NVDA
2. Start LibreOffice Writer
3. open the options dialog: "Tools" -> "Options", go to the "User Data" page
4. move focus to the "Company" text edit
5. open NVDA's Python console (Ctrl+Insert+Z)
6. try to print the object that the currently focused edit is labelled by:

```
>>> focus.labeledBy
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "C:\tools\cygwin\home\user\development\git\nvda\source\baseObject.py", line 59, in __get__
    return instance._getPropertyViaCache(self.fget)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\tools\cygwin\home\user\development\git\nvda\source\baseObject.py", line 167, in _getPropertyViaCache
    val = getterMethod(self)
          ^^^^^^^^^^^^^^^^^^
  File "C:\tools\cygwin\home\user\development\git\nvda\source\NVDAObjects\IAccessible\__init__.py", line 1167, in _get_labeledBy
    (pacc, accChild) = IAccessibleHandler.accNavigate(
    ^^^^^^^^^^^^^^^^
TypeError: cannot unpack non-iterable NoneType object

```

### Actual behavior:

An error occurs, s. above

### Expected behavior:

The label that is referenced by the IA2_RELATION_LABELLED_BY relation should be returned/printed in the Python console.

#### NVDA installed/portable/running from source:

running from source

#### NVDA version:

git master as of commit c32e0e78a7a627a800fbf58ca3a4eade9c3c4477

#### Windows version:

Windows 10 22H2

#### Name and version of other software in use when reproducing the issue:

LibreOffice 

Version: 25.2.0.0.alpha1+ (X86_64) / LibreOffice Community
Build ID: 4049405dcde61fa66852dae72d811562c69b3920
CPU threads: 12; OS: Windows 10 X86_64 (10.0 build 19045); UI render: default; VCL: win
Locale: en-GB (en_GB); UI: en-US
Calc: threaded

#### Other information about your system:

-

#### Does the issue still occur after restarting your computer?

n/a

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

NVDA 2024.3, same behavior

#### If NVDA add-ons are disabled, is your problem still occurring?

n/a

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?

n/a

### PR DESCRIPTION
**Pull request by michaelweghorn**
### Link to issue number:

Fixes #17436

### Summary of the issue:

IAccessible2 has an IA2_RELATION_LABELLED_BY relation type, see https://accessibility.linuxfoundation.org/a11yspecs/ia2/docs/html/group__grp_relations.html#ga7bbace7ffb57476b75d621af2e27d1ff . However, that one was not taken into account by NVDA's "labeledBy" property for objects. This could could be seen for example with LibreOffice that implements handling of that IAccessible2 relation.

Additionally, a `None` return value from
IAccessibleHandler.accNavigate wasn't handled
in `IAccessible._get_labeledBy`, triggering an error, e.g.:

    >>> focus.labeledBy
    Traceback (most recent call last):
      File "<console>", line 1, in <module>
      File "C:\tools\cygwin\home\user\development\git\nvda\source\baseObject.py", line 59, in __get__
        return instance._getPropertyViaCache(self.fget)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "C:\tools\cygwin\home\user\development\git\nvda\source\baseObject.py", line 167, in _getPropertyViaCache
        val = getterMethod(self)
              ^^^^^^^^^^^^^^^^^^
      File "C:\tools\cygwin\home\user\development\git\nvda\source\NVDAObjects\IAccessible\__init__.py", line 1167, in _get_labeledBy
        (pacc, accChild) = IAccessibleHandler.accNavigate(
        ^^^^^^^^^^^^^^^^
    TypeError: cannot unpack non-iterable NoneType object

### Description of user facing changes

None, unless they're using the Python console or addons making use of the property.

### Description of development approach

* Extend `IAccessibleHandler.RelationType` with the `LABELLED_BY` relation as defined in the IAccessible2 spec.
* Use this relation in `IAccessible._get_labeledBy` before falling back to trying a custom Mozilla/Gecko specific NAVRELATION.
* Add handling for the case where `IAccessibleHandler.accNavigate` for the custom Mozilla/Geck approach returns `None` to fix the error triggered otherwise when no relation is set.

### Testing strategy:

1. start NVDA
2. Start LibreOffice Writer
3. open the options dialog: "Tools" -> "Options", go to the "User Data" page
4. move focus to the "Company" text edit
5. open NVDA's Python console (Ctrl+Insert+Z)
6. print the object that the currently focused edit is labelled by, and it's accessible name and role:

    >>> focus.labeledBy
    <NVDAObjects.IAccessible.IAccessible object at 0x0BC55A70>
    >>> focus.labeledBy.name
    'Company:'
    >>> focus.labeledBy.role
    <Role.STATICTEXT: 7>

### Known issues with pull request:

None

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

@coderabbitai summary

### CODE REVIEW

**Review by LeonarddeR — APPROVED** [review]
Awesome!

File: source/NVDAObjects/IAccessible/__init__.py [hunk 1]
```diff
@@ -1163,12 +1163,19 @@ def isPointInObject(self, x, y):
 		return True
 
 	def _get_labeledBy(self):
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/__init__.py:1165]: please add a return type

```suggestion
	def _get_labeledBy(self) -> IAccessible | None:
```

File: user_docs/en/changes.md [hunk 2]
```diff
@@ -70,6 +70,7 @@ Specifically, MathML inside of span and other elements that have the attribute `
 * Opening the NVDA Python Console will no longer fail in case an error occurs while retrieving snapshot variables. (#17391, @CyrilleB79)
 * In Notepad and other UIA documents on Windows 11, if the last line is empty, the `braille next line command` will move the cursor to the last line.
 In any document, if the cursor is on the last line, it will be moved to the end when using this command. (#17251, @nvdaes)
+* In NVDA's Python console, retrieving the "labeledBy" property now works for objects in applications implementing the "labelled-by" IAccessible2 relation and no longer triggers an error. (#17436, @michaelweghorn)
```
Comment (seanbudd) [inline user_docs/en/changes.md:73]: This isn't a user facing change, if there's user impact, can you rephrase this to describe it. If there's no direct user impact, please move this to "changes for developers"

Comment (michaelweghorn) [reply user_docs/en/changes.md]: Done

File: source/NVDAObjects/IAccessible/__init__.py [hunk 3]
```diff
@@ -1947,7 +1947,7 @@ def _get_detailsRelations(self) -> Tuple["IAccessible"]:
 		# due to caching of baseObject.AutoPropertyObject, do not attempt to return a generator.
 		return tuple(detailsRelsGen)
 
-	def _get_controllerFor(self) -> List[NVDAObject]:
+	def _get_controllerFor(self) -> typing.List[NVDAObject]:
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/__init__.py:1950]:
```suggestion
	def _get_controllerFor(self) -> list[NVDAObject]:
```

File: source/NVDAObjects/IAccessible/__init__.py [hunk 4]
```diff
@@ -3,13 +3,13 @@
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
 
+from __future__ import annotations
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/__init__.py:6]: could you comment what this is needed for?

File: user_docs/en/changes.md [hunk 5]
```diff
@@ -100,6 +99,7 @@ Add-ons will need to be re-tested and have their manifest updated.
 * Removed the requirement to indent function parameter lists by two tabs from NVDA's Coding Standards, to be compatible with modern automatic linting. (#17126, @XLTechie)
 * Added the [VS Code workspace configuration for NVDA](https://nvaccess.org/nvaccess/vscode-nvda) as a git submodule. (#17003)
 * In the `brailleTables` module, a `getDefaultTableForCurrentLang` function has been added (#17222, @nvdaes)
+* Retrieving the "labeledBy" property now works for objects in applications implementing the "labelled-by" IAccessible2 relation and no longer triggers an error. (#17436, @michaelweghorn)
```
Comment (seanbudd) [inline user_docs/en/changes.md:102]:
```suggestion
* Retrieving the `labeledBy` property now works for objects in applications implementing the `labelled-by` IAccessible2 relation and no longer triggers an error. (#17436, @michaelweghorn)
```

File: source/NVDAObjects/IAccessible/__init__.py (hunk 4 above)
Comment (michaelweghorn) [reply source/NVDAObjects/IAccessible/__init__.py:9]: I've added a comment above the import. Is it OK like that?

**Review by seanbudd — APPROVED** [review]

File: source/NVDAObjects/IAccessible/__init__.py [hunk 6]
```diff
@@ -3,13 +3,13 @@
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
 
+
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/__init__.py:6]:
```suggestion
```

File: source/NVDAObjects/IAccessible/__init__.py [hunk 7]
```diff
@@ -1162,13 +1161,20 @@ def isPointInObject(self, x, y):
 			return False
 		return True
 
-	def _get_labeledBy(self):
+	def _get_labeledBy(self) -> "IAccessible" | None:
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/__init__.py:1164]:
```suggestion
	def _get_labeledBy(self) -> "IAccessible | None":
```

**Comment by michaelweghorn** [comment]
Thanks for the review and bringing this in line with the NVDA code conventions, @seanbudd !
